[JENKINS-42243] Increase functional test coverage

varyvol opened the pull request:
[JENKINS-42243](https://issues.jenkins-ci.org/browse/JENKINS-42243)

Increase coverage for SCM API events reacting to GH updates.

@reviewbybees @stephenc

stephenc commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:
```diff
@@ -27,6 +27,7 @@
 import com.cloudbees.jenkins.GitHubWebHook;
 import com.cloudbees.plugins.credentials.CredentialsNameProvider;
 import com.cloudbees.plugins.credentials.common.StandardCredentials;
+import com.google.common.annotations.VisibleForTesting;
```
Please do not use this annotation. In general avoid any dependency on either Guava or Google Commons in plugins. Those libraries do not respect the same backwards compatibility contract of Jenkins core.

varyvol replied:
@stephenc  I wasn't aware, thanks for pointing it out.

 Is it fine if I just add a comment in the method clarifying it is intended for testing purposes?

KostyaSha replied:
Why do you need annotate it at all if it in the same package?

varyvol replied:
Just for "visibility": that method exists so that it can be used in the tests to mock.

stephenc commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:
```diff
@@ -533,6 +534,11 @@ public void afterSave(@NonNull SCMNavigatorOwner owner) {
     }
 
+    @VisibleForTesting
+    GitHubWebHook getHook() {
```
unclear why you need to refactor this method. In any case you can probably just use `@Restricted(NoExternalUse.class)` rather than `@VisibleForTesting` even if you think you need this method

stephenc commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java:
```diff
@@ -1185,12 +1186,22 @@ protected boolean isCategoryEnabled(@NonNull SCMHeadCategory category) {
     }
 
+    @VisibleForTesting
+    SCMSourceOwner getOwnerNotSynchronized() {
```
Unclear what the requirement for this refactoring was

varyvol replied:
Just for testing purposes: allows mocking GitHubSCMSource and return some specific directly.

stephenc commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java:
```diff
@@ -1185,12 +1186,22 @@ protected boolean isCategoryEnabled(@NonNull SCMHeadCategory category) {
+    }
+
+    @VisibleForTesting
+    GitHubWebHook getHook() {
```
This just seems to be ad bad smell if you need to litter `getHook` methods in each class

stephenc commented on src/test/java/org/jenkinsci/plugins/github_branch_source/EventsTest.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    private void waitAndAssertReceived(boolean received) throws InterruptedException {
+        Thread.sleep(5100);
```
Oh dear me. No. Please see `SCMEvents.getWatermark()` and `SCMEvents.awaitAll(watermark)` in the scm-api tests artifact

varyvol replied:
Changed to this API. Although note that the 5 seconds delay is because the subscribers apply that delay before firing the event.

stephenc commented on src/test/java/org/jenkinsci/plugins/github_branch_source/WebhookTest.java:
```diff
@@ -0,0 +1,88 @@
+    private static String EXCEPTION_MESSAGE = "custom exception message";
+
+    @Test
+    public void registerHook() throws Exception {
```
Consider naming test cases using `given_X_when_Y_then_Z` pattern. For a smoke test like this seems to be just call it `smokes` or `methodName_smokes`

varyvol replied:
Sounds good.

stephenc commented on src/test/java/org/jenkinsci/plugins/github_branch_source/WebhookTest.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private void checkNavigator() {
+        GitHubSCMNavigator navigator = mock(GitHubSCMNavigator.class, Mockito.CALLS_REAL_METHODS);
```
These are very mock heavy tests and likely not useful other than to ensure that the sequence tested is not changed. As such it would seem to restrict refactoring rather than prevent bugs.

varyvol replied:
The idea is to verify that saving the navigator/source triggers a web hook registration.
I think it's a non-heavy test that provides confidence on that.

stephenc commented on src/test/java/org/jenkinsci/plugins/github_branch_source/EventsTest.java:
```diff
@@ -0,0 +1,174 @@
+    private static GHSubscriberEvent ghEvent;
+
+    @Test
+    public void pushEvents() throws Exception {
```
you are using a `@ClassRule` please split each test into individual test cases rather than piling on a load of scenarios in one test. Also name each test with what it is testing and what the outcome should be

varyvol replied:
Will do.

stephenc commented on src/test/java/org/jenkinsci/plugins/github_branch_source/EventsTest.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    private GHSubscriberEvent callOnEvent(PushGHEventSubscriber subscriber, String eventPayloadFile) throws IOException {
+        GHSubscriberEvent event = createMockEvent(eventPayloadFile);
```
Better to either use `SCMEvent.fireNow(createMockEvent)` given that you have registered as a test extension or submit the payload to the webhook endpoint directly, e.g. `ExtensionList.lookup(RootAction.class).get(com.cloudbees.jenkins.GitHubWebHook.class).doIndex(GHEvent, payload)` otherwise you are not really testing the functionality

varyvol replied:
@stephenc I am not sure of getting you here. I don't pretend to test the webhook but that a subscriber, once an event from github is received, analyses and processes it accordingly and uses the correct information to fire an SCMEvent.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:
```diff
@@ -533,6 +532,11 @@ public void afterSave(@NonNull SCMNavigatorOwner owner) {
         }
     }
 
+    @Restricted(NoExternalUse.class)
```
Such a restriction is pointless on a method which was already package access.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java:
```diff
@@ -1185,12 +1182,22 @@ protected boolean isCategoryEnabled(@NonNull SCMHeadCategory category) {
         }
     }
 
+    @Restricted(NoExternalUse.class)
```
ditto

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java:
```diff
@@ -1185,12 +1182,22 @@ protected boolean isCategoryEnabled(@NonNull SCMHeadCategory category) {
+        return getOwner();
+    }
+
+    @Restricted(NoExternalUse.class)
```
ditto

jglick commented on src/test/java/org/jenkinsci/plugins/github_branch_source/EventsTest.java:
```diff
@@ -0,0 +1,219 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
```
My general experience with Mockito/Powermock tests in Jenkins is that they are more trouble than they are worth. All sorts of internal refactorings with no effect on visible behavior, or Java APIs (where applicable—this plugin really should be exposing none), tend to make the tests need to be heavily edited to keep up.

Can we not extend the WireMock tests, which give us a far higher confidence that fixes are actually effective using the real behavior of GitHub, and which are not sensitive to purely internal changes?

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

varyvol commented:
@stephenc @jglick I have tried to address your comments.

I have removed the use of Mockito in EventsTest where it was not actually necessary, as the GHSubscriberEvent could be created directly through a constructor. For the Webhooks test I have left it there, as I think it's an API that won't change usually and I was not able to think in an alternative. If you can think in a better way of testing it or you think it's preferable not to have them, just let me know.

I have also modified the tests to reduce the events delay so that they are faster.

varyvol commented:
@stephenc @jglick did you have the opportunity to review the latest changes?

stephenc commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:
```diff
@@ -598,6 +597,10 @@ public void afterSave(@NonNull SCMNavigatorOwner owner) {
         }
     }
 
+    GitHubWebHook getHook() {
```
:-1: I am most likely going to refactor the webhook code anyway, very risky to use mockito for this

varyvol replied:
@stephenc are you fine with the events test? If so, I can remove the webhook stuff to push this forward and, when you later refactor the webhook code, I can review that part.

stephenc replied:
The event tests that are not using mockito... those seemed ok. everything else is not interesting to my eyes

varyvol replied:
@stephenc let's keep with them then.

varyvol commented:
@stephenc I think GitHub allows doing it when merging?

stephenc commented:
That means doing other than a merge commit, which is not the style of history I want. Please squash manually

varyvol commented:
@stephenc I messed it up with this one, so I have created [PR #132](https://github.com/jenkinsci/github-branch-source-plugin/pull/132).
